File: app/src/main/java/com/example/toshiba/parkme/ProfileActivity.java
package com.example.toshiba.parkme;

import android.app.Activity;
import android.content.Intent;
import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;
import android.view.View;
import android.view.View.OnClickListener;
import android.widget.ArrayAdapter;
import android.widget.Button;
import android.widget.TextView;
import android.widget.Toast;

import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.auth.FirebaseUser;
import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseError;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;
import com.google.firebase.database.ValueEventListener;

import java.util.Calendar;


public class ProfileActivity extends AppCompatActivity {
//
//    ArrayAdapter<String> adapter;
    private Button list_button;
    private Button signout_button,status;
    private TextView Thename;


    //firebase
   FirebaseUser Current_user;
   DatabaseReference databaseReference1,dref1;

   int f,t,loc;
   int avv,bkk;

   //loc variable 1 ==SAM   ;2==Sd 3== RC

//    @Override
//    protected void onStart() {
//        super.onStart();
//
//
//    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_profile);

        Thename = (TextView) findViewById(R.id.nameid);
        Current_user = FirebaseAuth.getInstance().getCurrentUser();
        // if(Current_user != null) {
        databaseReference1 = FirebaseDatabase.getInstance().getReference();
        // }
 //       Toast.makeText(getApplicationContext(),"current"+Current_user.getUid(),Toast.LENGTH_LONG).show();

        dref1 = FirebaseDatabase.getInstance().getReference();

        databaseReference1.addValueEventListener(new ValueEventListener() {
            @Override
            public void onDataChange(DataSnapshot dataSnapshot) {
                //itemlist = new ArrayList<>();
                //                //String user_name=dataSnapshot.child(uid).child("name").getValue(String.class);
                //                //User user = dataSnapshot.getValue(User.class);
                //                //Thename.setText(user.setName());
                String name = dataSnapshot.child("Users").child(Current_user.getUid()).child("name").getValue(String.class);
                name= name.toUpperCase();
                Thename.setText(name);
                f =dataSnapshot.child("Users").child(Current_user.getUid()).child("from").getValue(Integer.class);
                t =dataSnapshot.child("Users").child(Current_user.getUid()).child("to").getValue(Integer.class);
                loc = dataSnapshot.child("Users").child(Current_user.getUid()).child("loc").getValue(Integer.class);
                if(loc == 1)
                {
                    avv = dataSnapshot.child("Parking Area").child("SAM").child("Available").getValue(Integer.class);
                    bkk = dataSnapshot.child("Parking Area").child("SAM").child("Booked").getValue(Integer.class);
                }else if (loc ==2)
                {
                    avv = dataSnapshot.child("Parking Area").child("Sd").child("Available").getValue(Integer.class);
                    bkk = dataSnapshot.child("Parking Area").child("Sd").child("Booked").getValue(Integer.class);
                } else if (loc ==3)
                {
                    avv = dataSnapshot.child("Parking Area").child("RC").child("Available").getValue(Integer.class);
                    bkk = dataSnapshot.child("Parking Area").child("RC").child("Booked").getValue(Integer.class);
                }
                //           Toast.makeText(getApplicationContext(),"Fr Too"+f+t,Toast.LENGTH_LONG).show();

                // f = Integer.parseInt(fr);
                // t = Integer.parseInt(too);

                //hourly refresh
                Calendar cal = Calendar.getInstance();
                //final int minute = cal.get(Calendar.MINUTE);
                //12 hour format
                //int hour = cal.get(Calendar.HOUR);
                //24 hour format
                final int hourofday = cal.get(Calendar.HOUR_OF_DAY);
                if(t <=hourofday && (t!=0)) {

                    if (loc == 1) {
                        dref1.child("Users").child(Current_user.getUid()).child("from").setValue(0);
                        dref1.child("Users").child(Current_user.getUid()).child("to").setValue(0);
                        dref1.child("Users").child(Current_user.getUid()).child("loc").setValue(0);
                        dref1.child("Parking Area").child("SAM").child("Available").setValue(avv + 1);
                        dref1.child("Parking Area").child("SAM").child("Booked").setValue(bkk - 1);
                    } else if (loc == 2) {
                        dref1.child("Users").child(Current_user.getUid()).child("from").setValue(0);
                        dref1.child("Users").child(Current_user.getUid()).child("to").setValue(0);
                        dref1.child("Users").child(Current_user.getUid()).child("loc").setValue(0);
                        dref1.child("Parking Area").child("Sd").child("Available").setValue(avv + 1);
                        dref1.child("Parking Area").child("Sd").child("Booked").setValue(bkk - 1);
                    } else if (loc == 3) {
                        dref1.child("Users").child(Current_user.getUid()).child("from").setValue(0);
                        dref1.child("Users").child(Current_user.getUid()).child("to").setValue(0);
                        dref1.child("Users").child(Current_user.getUid()).child("loc").setValue(0);
                        dref1.child("Parking Area").child("RC").child("Available").setValue(avv + 1);
                        dref1.child("Parking Area").child("RC").child("Booked").setValue(bkk - 1);
                    }

                }


            }

            @Override
            public void onCancelled(DatabaseError databaseError) {

            }
        });



//        if(f==-1 && t==-1)
//        {
//            if(loc == 1)
//            {
//
//                dref1.child("Parking Area").child("SAM").child("Available").setValue(avv + 1);
//                dref1.child("Parking Area").child("SAM").child("Booked").setValue(bkk - 1);
//            }else if (loc ==2)
//            {
//                dref1.child("Parking Area").child("SAM").child("Available").setValue(avv + 1);
//                dref1.child("Parking Area").child("SAM").child("Booked").setValue(bkk - 1);
//            } else if (loc ==3) {
//                dref1.child("Parking Area").child("SAM").child("Available").setValue(avv + 1);
//                dref1.child("Parking Area").child("SAM").child("Booked").setValue(bkk - 1);
//            }
//            f=0;
//            t=0;
//        }


        list_button = (Button) findViewById(R.id.list);
        signout_button = (Button) findViewById(R.id.logout_button);
        status = (Button) findViewById(R.id.statusbutton);

        list_button.setOnClickListener((new OnClickListener() {
            @Override
            public void onClick(View view) {
                Intent listIntent = new Intent(ProfileActivity.this, BookActivity.class);
                if(!(f==0 && t==0)){
                   Toast.makeText(getApplicationContext(),"ALREADY BOOKED! CHECK STATUS",Toast.LENGTH_LONG).show();
               }else {
                    startActivity(listIntent);
                }
            }
        }));

        status.setOnClickListener(new OnClickListener() {
            @Override
            public void onClick(View v) {
                Intent Intent = new Intent(ProfileActivity.this, StatusShow.class);
                startActivity(Intent);
            }
        });
        signout_button.setOnClickListener((new OnClickListener() {
            @Override
            public void onClick(View view) {

                //Log out action
                FirebaseAuth.getInstance().signOut();
                Toast.makeText(getApplicationContext(),"SIGNED OUT",Toast.LENGTH_LONG).show();
                Intent listIntent = new Intent(ProfileActivity.this, MainActivity.class);

                startActivity(listIntent);
            }
        }));

    }

    @Override
    public void onBackPressed()
    {
        FirebaseAuth.getInstance().signOut();
      //  Intent listIntent = new Intent(ProfileActivity.this, MainActivity.class);
      //  startActivity(listIntent);
        finish();

    }

}

File: app/src/main/java/com/example/toshiba/parkme/activity/ProfileActivity.kt
package com.example.toshiba.parkme.activity

import android.content.Context
import android.content.DialogInterface
import android.content.Intent
import android.content.SharedPreferences.Editor
import android.os.Bundle
import android.view.View
import android.widget.Button
import android.widget.ProgressBar
import android.widget.TextView
import android.widget.Toast
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import com.example.toshiba.parkme.Constants
import com.example.toshiba.parkme.Constants.loginSharedPref
import com.example.toshiba.parkme.R
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.auth.FirebaseUser
import com.google.firebase.database.*
import java.util.*


class ProfileActivity : AppCompatActivity() {

    private var list_button: Button? = null
    private var signout_button: Button? = null
    private var status: Button? = null
    private var thename: TextView? = null
    private var progressDialog: ProgressBar? = null

    private var currentUser: FirebaseUser? = null
    var databaseReference: DatabaseReference? = null
    var from = 0
    var to = 0
    var loc = 0
    var avv = 0
    var bkk = 0

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_profile)
        thename = findViewById(R.id.nameid)
        currentUser = FirebaseAuth.getInstance().currentUser
        databaseReference = FirebaseDatabase.getInstance().reference
        progressDialog = findViewById(R.id.progressbar)
        list_button = findViewById<View>(R.id.list) as Button
        signout_button = findViewById<View>(R.id.logout_button) as Button
        status = findViewById<View>(R.id.statusbutton) as Button
        list_button!!.setOnClickListener {
            val listIntent = Intent(this@ProfileActivity, BookActivity::class.java)
            if (!(from == 0 && to == 0)) {
                Toast.makeText(applicationContext, "ALREADY BOOKED! CHECK STATUS", Toast.LENGTH_LONG).show()
            } else {
                startActivity(listIntent)
            }
        }
        status!!.setOnClickListener {
            val intent = Intent(this@ProfileActivity, StatusActivity::class.java)
            startActivity(intent)
        }
        signout_button!!.setOnClickListener { //Log out action
            val builder: AlertDialog.Builder = AlertDialog.Builder(this)
            builder.setMessage("If you log out, your booking will be cancelled, if any. Are you sure you want to proceed?").setTitle("WARNING!")
                    .setCancelable(false)
                    .setPositiveButton("Yes, Log me out", DialogInterface.OnClickListener { _, _ ->
                        FirebaseAuth.getInstance().signOut()
                        val prefs = getSharedPreferences(loginSharedPref, Context.MODE_PRIVATE)
                        val edit: Editor = prefs.edit()
                        edit.clear()
                        edit.putString(Constants.userLoginStatus, Constants.isNotLoggedIn)
                        edit.apply()
                        resetParams()
                        finish()
                        Toast.makeText(applicationContext, "SIGNED OUT", Toast.LENGTH_LONG).show()
                        val listIntent = Intent(this@ProfileActivity, MainActivity::class.java)
                        startActivity(listIntent)
                    })
                    .setNegativeButton("No, I'll stay") { dialog, _ ->
                        dialog.cancel()
                    }
            val alertDialog: AlertDialog = builder.create()
            alertDialog.show()
        }
    }

    override fun onBackPressed() {
        FirebaseAuth.getInstance().signOut()
        finish()
    }

    override fun onResume() {
        super.onResume()
        databaseReference!!.addValueEventListener(object : ValueEventListener {
            override fun onDataChange(dataSnapshot: DataSnapshot) {
                var name = dataSnapshot.child("Users").child(currentUser!!.uid).child("name").getValue(String::class.java)!!
                name = name.toUpperCase(Locale.getDefault())
                thename!!.text = name
                progressDialog?.visibility = View.GONE
                from = dataSnapshot.child("Users").child(currentUser!!.uid).child("from").getValue(Int::class.java)!!
                to = dataSnapshot.child("Users").child(currentUser!!.uid).child("to").getValue(Int::class.java)!!
                loc = dataSnapshot.child("Users").child(currentUser!!.uid).child("loc").getValue(Int::class.java)!!
                if (loc == 1) {
                    avv = dataSnapshot.child("Parking Area").child("SAM").child("Available").getValue(Int::class.java)!!
                    bkk = dataSnapshot.child("Parking Area").child("SAM").child("Booked").getValue(Int::class.java)!!
                } else if (loc == 2) {
                    avv = dataSnapshot.child("Parking Area").child("Sd").child("Available").getValue(Int::class.java)!!
                    bkk = dataSnapshot.child("Parking Area").child("Sd").child("Booked").getValue(Int::class.java)!!
                } else if (loc == 3) {
                    avv = dataSnapshot.child("Parking Area").child("RC").child("Available").getValue(Int::class.java)!!
                    bkk = dataSnapshot.child("Parking Area").child("RC").child("Booked").getValue(Int::class.java)!!
                }

                updateStatus()

            }

            override fun onCancelled(databaseError: DatabaseError) {
                progressDialog?.visibility = View.GONE
                //error occurred
            }
        })
    }

    private fun resetParams() {
        when (loc) {
            1 -> {
                databaseReference!!.child("Users").child(currentUser!!.uid).child("from").setValue(0)
                databaseReference!!.child("Users").child(currentUser!!.uid).child("to").setValue(0)
                databaseReference!!.child("Users").child(currentUser!!.uid).child("loc").setValue(0)
                databaseReference!!.child("Parking Area").child("SAM").child("Available").setValue(avv + 1)
                databaseReference!!.child("Parking Area").child("SAM").child("Booked").setValue(bkk - 1)
            }
            2 -> {
                databaseReference!!.child("Users").child(currentUser!!.uid).child("from").setValue(0)
                databaseReference!!.child("Users").child(currentUser!!.uid).child("to").setValue(0)
                databaseReference!!.child("Users").child(currentUser!!.uid).child("loc").setValue(0)
                databaseReference!!.child("Parking Area").child("RC").child("Available").setValue(avv + 1)
                databaseReference!!.child("Parking Area").child("RC").child("Booked").setValue(bkk - 1)
            }
            3 -> {
                databaseReference!!.child("Users").child(currentUser!!.uid).child("from").setValue(0)
                databaseReference!!.child("Users").child(currentUser!!.uid).child("to").setValue(0)
                databaseReference!!.child("Users").child(currentUser!!.uid).child("loc").setValue(0)
                databaseReference!!.child("Parking Area").child("Sd").child("Available").setValue(avv + 1)
                databaseReference!!.child("Parking Area").child("Sd").child("Booked").setValue(bkk - 1)
            }
        }
    }

    private fun updateStatus() {
        // for a commercial model, the available and booked numbers for each Parking Area should be updated in backend
        // this method (runs once every time Activity is opened) will update the user values, although it should also be handled in backend
        //TODO store to, avv, bkk in sharedPref
        val cal = Calendar.getInstance()
        val hourOfDay = cal[Calendar.HOUR_OF_DAY]
        if (to <= hourOfDay && to != 0) {
            resetParams()
        }
    }
}